platform/cxxsupport/mstd_span

```diff
@@ -0,0 +1,381 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MSTD_SPAN_
+#define MSTD_SPAN_
+
+#if __cplusplus >= 201703L && __has_include(<span>)
+#include <version>
+#endif
+
+#if __cpp_lib_span >= 202002L
+#include <span>
+
+namespace mstd {
+    using std::span;
+}
+
+#else //__cpp_lib_span >= 202002L
+
+#include <array>
+#include <mstd_iterator>
+#include <mstd_type_traits>
+#include "mbed_assert.h"
+
+namespace mstd {
+
+constexpr size_t dynamic_extent = -1;
+
+template <typename, size_t = dynamic_extent> class span;
+
+namespace detail {
+
+template<typename ElementType, size_t Extent>
+class storage {
```

Comment:
I think I'd just make this a `struct` - there's no encapsulation magic here, given you're just accessing public members.